cache-explainer.md

```diff
@@ -0,0 +1,104 @@
+
+This proposal documents ongoing discussions in the W3C WebML Working Group and builds on existing mechanisms in frameworks like ONNX Runtime.
+
+## Goals
+
+- Provide a mechanism for web applications to save and load compiled `MLGraph` objects.
+- Reduce the time required for repeated ML model inference by avoiding redundant graph compilation.
+- Ensure compatibility with existing WebNN API constructs and workflows.
+
+## Non-goals
+
+- This proposal does not aim to define a universal format for graph serialization across all frameworks.
+- It does not address caching mechanisms for non-WebNN APIs or other types of computational graphs.
+- Cross-origin model sharing is out of scope.
+
+## User research
+
+[If any user research has been conducted to inform the design choices presented,
+discuss the process and findings.
+We strongly encourage that API designers consider conducting user research to
+verify that their designs meet user needs and iterate on them,
+though we understand this is not always feasible.]
+
+## Use cases
+
+### Reduce time to first inference
+
+A web application performing real-time image recognition can save the compiled graph after the first inference. If the page is reloaded, subsequent inferences reuse the cached graph, significantly reducing latency by avoiding both the model redownload and recompilation steps.
+
+## Proposed API
+
+```webidl
+partial interface MLContext {
+  Promise<sequence<DOMString>> listGraphs();
+  Promise<MLGraph> loadGraph(DOMString key);
+  Promise<undefined> saveGraph(DOMString key, MLGraph graph);
+  undefined deleteGraph(DOMString key);
+};
+```
+
+1. **`listGraphs()`**: Returns a list of keys for all cached graphs.
+2. **`loadGraph(key)`**: Loads a cached graph associated with the given key.
+3. **`saveGraph(key, graph)`**: Saves the provided graph under the specified key.
+4. **`deleteGraph(key)`**: Deletes the cached graph associated with the given key.
+
+## Considered alternatives
+
+### Explicit vs implicit API
+
+>GPU shader caching is implicit, however the difference is that a shader program is a small input and so it's easy for the site to regenerate the shader so the browser can hash it to compare with the cache. ML models on the other hand are large because of the weights. Loading all the weights just to discover that a cached version of the model is available would be a waste of time and resources. (via [comment](https://github.com/webmachinelearning/webnn/issues/807#issuecomment-2608135598))
```

Comment:
Note as well that the model can't be compiled without the weights because the implementation may perform device-specific constant folding and memory layout optimizations.

Reply:
Fixed in https://github.com/webmachinelearning/webnn/pull/862/commits/20ed41475d6cd99ce793ebf76d19baa7275e419c